### ISSUE TITLE
pub global activate via path not working unless `.dart_tool/pub/bin/<executable>` is deleted

### ISSUE THREAD

**Issue by btrautmann** [dart-lang/pub#4409]
```
#### General info

- Dart 3.5.0 (stable) (Tue Jul 30 02:17:59 2024 -0700) on "macos_arm64"
- on macos / Version 14.7 (Build 23H124)
- locale is en-US

#### Project info

- sdk constraint: '^3.5.0'
- dependencies: 
- dev_dependencies: lints, test

#### Process info

|  Memory |  CPU | Elapsed time | Command line                                                                    |
| ------: | ---: | -----------: | ------------------------------------------------------------------------------- |
|   23 MB | 0.0% |     22:57:26 | dart devtools --machine --allow-embedding --dtd-uri ws:<path>/SLOui8f56EyIjsj7  |
| 2278 MB | 0.0% |     22:57:27 | dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.96.0 |
|  464 MB | 0.0% |        03:18 | dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.96.0 |
|   22 MB | 0.0% |     22:57:27 | dart tooling-daemon --machine                                                   |
|   82 MB | 0.0% |        03:18 | dart tooling-daemon --machine                                                   |
|   49 MB | 0.3% |     22:57:27 | flutter_tools.snapshot daemon                                                   |
```

## The Issue

I started noticing a few weeks ago that `dart pub global activate --source="path" <path>` stopped working. To reproduce, you can:

1. use a project created via `dart create -t console <name>`
2. add the following to the `pubspec.yaml`
```yaml
executables:
  name: name
```
3. activate the executable globally via `dart pub global activate --source="path" <path/to/root/of/project>`
4. Run the executable via `name` and note result
5. Change the logic of the executable (in the case of the `dart create -t console` default project, update `calculate()` to return a different integer).
6. Repeat step 3-4, noting that the result has not changed. 
7. Delete the `.dart_tool/pub/bin/<name>` directory
8. Repeat steps 3-4, noting that the result does change this time.

[Here is a sample repo](https://github.com/btrautmann/global_activate_repro) you can clone and use to do the above.

### Expected results

Prior to this bug, re-running `dart pub global activate --source="path"` would correctly update the activated executable such that the new logic would be respected.

**Comment by sigurdm** [comment]
Yeah - this is a bug. We are using snapshots, but we don't get to invalidate them (as dart run does).

**Comment by sigurdm** [comment]
Perhaps best would be to have the binstub for a path-activated package just call `dart path/to/package/bin/executable.dart` instead of precompiling.

Alternatively the binstub needs to recompile (like we do in `dart run`).

**Comment by btrautmann** [comment]
> Yeah - this is a bug. We are using snapshots, but we don't get to invalidate them (as dart run does).

Just curious: This at one point _did_ work the way I expected; did something change?

**Comment by sigurdm** [comment]
A bisect shows it was introduced by a3ee209988af39c1d4f3c67b4910c317047356ca #3586

I can see that the claim "Always generate the full logic in the binstub (it will check for snapshot presence anyways)" is missing the point. There is no check if the code is updated relative to the snapshot.

**Comment by sigurdm** [comment]
Sorry about that. We should probably revert to having the binstubs just invoke `dart global run <dart-file>.dart` instead of trying to use a snapshot.

**Comment by SandPod** [comment]
Is this a duplicate of https://github.com/dart-lang/pub/issues/4295?

**Comment by btrautmann** [comment]
Ironically I ran into this while repro'ing another [bug](https://github.com/dart-lang/pub/issues/4536)... Any idea when this might be able to be picked up?

**Comment by javaddict** [comment]
There’s another bug here. Even after you delete the snapshot, repeat steps 3–4, and obtain the updated results, running `<name>` still doesn’t behave as you expect. You’ll notice that now every time you run `<name>`, it rebuilds and runs, instead of using the new snapshot you just regenerated by deleting the old one. This happens because step 4 scrambles the snapshot path inside `<name>` (Actually, step 3 has no effect on regenerating the snapshot). However, if you run step 3 again, it will write the correct snapshot path back into `<name>`, and from then on, each execution of `<name>` works as all we expect.
